source/initial_temperature/patch_on_S40RTS.cc

```diff
@@ -0,0 +1,197 @@
+#include <aspect/utilities.h>
+
+namespace aspect
+{
+  namespace InitialTemperature
+  {
+    template <int dim>
+    PatchOnS40RTS<dim>::PatchOnS40RTS ()
+    {}
+
+
+    template <int dim>
+    void
+    PatchOnS40RTS<dim>::initialize ()
+    {
+      Utilities::AsciiDataInitial<dim>::initialize(1);
+      if (s40rts.vs_to_density_method == s40rts.file)
+        {
+          s40rts.profile.initialize(this->get_mpi_communicator());
+          s40rts.vs_to_density_index = s40rts.profile.get_column_index_from_name("vs_to_density");
+        }
+    }
+
+
+    template <int dim>
+    double
+    PatchOnS40RTS<dim>::
+    ascii_grid_vs (const Point<dim> &position) const
+    {
+      const double vs_perturbation = Utilities::AsciiDataInitial<dim>::get_data_component(position,0);
+      return vs_perturbation;
+    }
+
+
+    template <int dim>
+    double
+    PatchOnS40RTS<dim>::
+    initial_temperature (const Point<dim> &position) const
+    {
+      const double depth = this->get_geometry_model().depth(position);
+
+      double vs_perturbation;
+      if (depth <= max_grid_depth - smoothing_length_scale)
+        {
+          vs_perturbation = ascii_grid_vs(position);
+        }
+      //add smoothing between the two models
+      else if (depth > max_grid_depth - smoothing_length_scale && depth < max_grid_depth)
+        {
+          const double scale_factor = (depth-(max_grid_depth-smoothing_length_scale))/smoothing_length_scale;
+          vs_perturbation = s40rts.get_Vs(position)*(scale_factor) + ascii_grid_vs(position)*(1-scale_factor);
+        }
+      else
+        {
+          vs_perturbation = s40rts.get_Vs(position);
+        }
+
+
+      // use either the user-input reference temperature as background temperature
+      // (incompressible model) or the adiabatic temperature profile (compressible model)
+      const double background_temperature = this->get_material_model().is_compressible() ?
+                                            this->get_adiabatic_conditions().temperature(position) :
+                                            s40rts.reference_temperature;
+
+      // get the Vs to density conversion
+      double vs_to_density = 0.0;
+      if (s40rts.vs_to_density_method == s40rts.file)
+        vs_to_density = s40rts.profile.get_data_component(Point<1>(depth), s40rts.vs_to_density_index);
+      else if (s40rts.vs_to_density_method == s40rts.constant)
+        vs_to_density = s40rts.vs_to_density_constant;
+      else
+        // we shouldn't get here but instead should already have been
+        // kicked out by the assertion in the parse_parameters()
+        // function
+        Assert (false, ExcNotImplemented());
+
+      // scale the perturbation in seismic velocity into a density perturbation
+      // vs_to_density is an input parameter
+      const double density_perturbation = vs_to_density * vs_perturbation;
+
+      double temperature_perturbation;
+      if (depth > s40rts.no_perturbation_depth)
```

Comment:
I think the `s40rts.no_perturbation_depth` should only work on the s40rts data. Now it could also remove the ascii grid data, which defeats the purpose of adding a upper mantle model. Maybe move this check to line 79 and set the vs_perturbation from s40rts to zero there when above the `s40rts.no_perturbation_depth`.

Then you don't need the conditional statements and just go for `const double temperature_perturbation = ...`

Do you actually need the `s40rts.no_perturbation_depth`? It is not used above `max_grid_depth-smoothing_length_scale` and quite strange if it lies below that depth.

Or do you need it because you would still like to describe the temperature in some upper part of the mantle in a different way than through any seismic velocity conversion (i.e. nor through the upper mantle model nor through s40rts)? In that case, I think it's more clear to add a separate `no_perturbation_depth` for this combined model, and to return the background_temperature for depths smaller than `no_perturbation_depth` early on.

Reply:
So we do actually want the s40rts.no_perturbation_depth to work on the ascii grid data too - the idea is that because the continental lithosphere has really high/low temperature values due to the Vs to temperature scalings we tend to use, we want to have the option to zero out the top 100 or 200 km in order to remove this bias. The upper mantle tomography is still useful as it goes down to 600/700 km.
We are thinking about better ways to handle the continents but this is an okay option for now.

Let me know what you think!